feat: save default live events columns

kappa90 opened the pull request:
## Problem

Fixes #10143

The user can now save live events columns as a project/team-wide default, as well as reorder them by drag and drop

## Changes

- Visible columns can be reordered by drag and drop
- Styled the modal according to [this Figma](https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=11161%3A49783) but keeping the columns stacked horizontally (@clarkus please check if anything needs to be refined)
- Added a `Save as default for all project members` button which, if toggled, saves the current choice to db

![image](https://user-images.githubusercontent.com/7335343/173567799-c620411a-3072-4c65-990d-1d46ca6102d0.png)

![Peek 2022-06-14 13-13](https://user-images.githubusercontent.com/7335343/173564896-6d1cfaae-5cf7-455d-9b46-8e96bebae0f8.gif)

## How did you test this code?

- Added unit tests to the columnConfiguratorLogic
- Added a cypress test to add and remove a column
- Did NOT test the drag and drop on cypress

kappa90 committed:
feat: save default live events columns

kappa90 committed:
Merge remote-tracking branch 'origin/master' into save-live-events-columns

kappa90 committed:
added column config logic test

kappa90 committed:
ran prettier

github-actions[bot] committed:
Update snapshots

kappa90 committed:
fix api.mock.ts

kappa90 committed:
Merge remote-tracking branch 'origin/save-live-events-columns' into save-live-events-columns

mariusandra commented on frontend/src/lib/components/ResizableTable/TableConfig.scss:
```diff
@@ -24,13 +23,65 @@
 
         &.selected {
-            background-color: var(--primary-hover);
+            background-color: #eef2ff;
```
Is this something that could go in a var? We should try to stick to standard and globally specified colors whenever possible.

kappa90 replied:
Done, although I think this is a new color that is in Figma but has not been implemented in the UI, I ask @clarkus for confirmation.

clarkus replied:
That's the highlight blue value we use in menus and other places. It should already exist in a variable somewhere. @Twixes would probably know for sure.

kappa90 replied:
It's only used here as far as I can see

![image](https://user-images.githubusercontent.com/7335343/173618627-932c000d-b011-42c8-a36f-1eb37aa7c6d2.png)

Twixes replied:
Indeed, `--recording-current-event` seems to be `--primary-bg-hover`, while `--recording-current-event-dark` is `--primary-alt`. Practically every color used in the app should be defined in `global.scss`, unless it _has_ to be generated programmatically.

mariusandra commented on frontend/src/lib/components/ResizableTable/TableConfig.scss:
```diff
@@ -24,13 +23,65 @@
+        .lists {
+            background-color: var(--bg-side);
+            border-radius: 4px;
+            padding: 8px;
```
Here and elsewhere: `8px` -> `0.5rem`

https://storybook.posthog.net/?path=/story/lemon-ui-overview--page#lemon-principles

mariusandra commented on frontend/src/lib/components/ResizableTable/TableConfig.scss:
```diff
@@ -24,13 +23,65 @@
+    }
+}
+
+.column-configurator-modal-sortable-container {
+    z-index: 9999;
+    display: flex;
+    align-items: center;
+    justify-content: flex-start;
+    padding: 0.5rem;
+
+    .drag-handle {
+        color: var(--text-default);
+        font-size: 1.2em;
+        padding-right: 5px;
+        transform: rotate(90deg) translateX(2px);
```
Can this be nested inside the other css class?

kappa90 replied:
Unfortunately this is a standalone helper class that gets added to the sortable preview during drag, so nesting it won't make it work.

kappa90 replied:
I am now investigating how to implement the drag and drop between the two columns.

kappa90 replied:
[React Sortable HOC](http://clauderic.github.io/react-sortable-hoc/#/basic-configuration/multiple-lists?_k=pzzyd2) does not support drag and drop between multiple lists, we'd have to implement it ourself, not trivial.

mariusandra reviewed (changes requested):
- I'm not sure why... but it doesn't save things for me:
![2022-06-14 15 15 31](https://user-images.githubusercontent.com/53387/173586346-8134586b-0321-4f87-adf7-2e1068043feb.gif)

Things don't end up in the database, and I don't even see the action to update the team in the redux devtools...

- It happened more than once that I tried to drag a column from the right list into the left list:
![2022-06-14 15 26 31](https://user-images.githubusercontent.com/53387/173588497-4e70df60-2708-4af2-baad-e49d269a3a81.gif)

Would that be hard to add? 🤔

kappa90 committed:
fixes after review

kappa90 committed:
fix test

github-actions[bot] committed:
Update snapshots

kappa90 committed:
Merge remote-tracking branch 'origin/master' into save-live-events-columns

clarkus reviewed (changes requested):
Looks like a great start. Here's my feedback:

- When an available column is selected, should it be removed from the available list and then only shown in the visible list? 
- I'm not sure we need the checkboxes for selecting items if we're maintaining separate lists. If we use them, I feel like they should also be shown in the available list to have a consistent way of selecting / deselecting items. Given we're using multiple columns, I'd be fine dropping the checkboxes entirely and using inline remove actions instead.
- The reserved / locked columns shouldn't be able to be reordered. If we show a different column first that would imply that we're sorting things differently and grouping differently. Feels like the primary event identifier should always be reserved for the first column / primary dimension. 
- When the save is executed, we should show a toast notification that communicates the result of the action - successfully saved or error. 
- Is the checkbox for saving as the default limited to only users with management or administrator rights? Just want to be sure that's not something available to every user. Speaking of, what happens if a user has customized their columns and then a new default is created by an admin. Does that override the custom setup for that individual user or is it unaffected until the user "resets to the defaults"?

Possible scope creep, but this table is sorted by time by default. We should indicate that in the table column using our standard sort indicators. Not saying we should allow users to change the sort, just saying that clarifying how sorting is applied would make it more obvious. This is a pre-existing issue, so don't think it's blocking for this PR. 

Nice job, @kappa90!

kappa90 commented:
> * When an available column is selected, should it be removed from the available list and then only shown in the visible list?
I spoke with @mariusandra about it and this is not so straightforward, it would require to change the taxonomic filter to call the backend requesting to exclude some results, it can't just be easily hidden on the frontend. I'd keep it out of scope.

> * I'm not sure we need the checkboxes for selecting items if we're maintaining separate lists. If we use them, I feel like they should also be shown in the available list to have a consistent way of selecting / deselecting items. Given we're using multiple columns, I'd be fine dropping the checkboxes entirely and using inline remove actions instead.

Removed the checkboxes.

> * The reserved / locked columns shouldn't be able to be reordered. If we show a different column first that would imply that we're sorting things differently and grouping differently. Feels like the primary event identifier should always be reserved for the first column / primary dimension.

Disabled columns can't be dragged and dropped now.

> * When the save is executed, we should show a toast notification that communicates the result of the action - successfully saved or error.

There is a toast notification already.

> * Is the checkbox for saving as the default limited to only users with management or administrator rights? Just want to be sure that's not something available to every user. Speaking of, what happens if a user has customized their columns and then a new default is created by an admin. Does that override the custom setup for that individual user or is it unaffected until the user "resets to the defaults"?
> 

Made it so that the button is only clickable by staff members.

I think that if the CI goes, it should be good to go.

![image](https://user-images.githubusercontent.com/7335343/173801134-1e845a5c-3047-4a1d-a218-a40f401992ad.png)

![Peek 2022-06-15 12-03](https://user-images.githubusercontent.com/7335343/173801673-e89c871c-f163-460d-9f3f-ffdc3fb3dca7.gif)

kappa90 committed:
fixes after review

mariusandra commented on posthog/models/team/team.py:
```diff
@@ -111,6 +111,7 @@ class Team(UUIDClassicModel):
     timezone: models.CharField = models.CharField(max_length=240, choices=TIMEZONES, default="UTC")
     data_attributes: models.JSONField = models.JSONField(default=get_default_data_attributes)
     person_display_name_properties: ArrayField = ArrayField(models.CharField(max_length=400), null=True, blank=True)
+    live_events_columns: ArrayField = ArrayField(models.CharField(max_length=200), null=True, blank=True)
```
Let's make it a `TextField`. Somebody might want to put three 60-character-long columns here, and boom, we're out of space.

mariusandra commented on frontend/src/lib/components/ResizableTable/TableConfig.scss:
```diff
@@ -24,7 +23,7 @@
 
         &.selected {
-            background-color: var(--primary-hover);
+            background-color: var(--bg-light-blue);
```
This should be `--primary-bg-hover`, like [said here](https://github.com/PostHog/posthog/pull/10293#discussion_r897004326). 

That color is defined in a roundabout way [here](https://github.com/PostHog/posthog/blob/fd11b357c09b76a50fd0c368cce5874966b57520/frontend/src/styles/global.scss#L19). It's actually a strong blue with an alpha channel, not a light blue :D.

mariusandra reviewed (commented):
1. Might be out of scope: The titles that are truncated with `...` aren't left-aligned

<img width="854" alt="image" src="https://user-images.githubusercontent.com/53387/173806115-ce33c168-1c5c-497c-81ed-19c1284f829a.png">

2. Just seeing those draggable boxes on the left makes me unconsciously want to drag the ones on the right. However then I select some text and mess up everything. 

![2022-06-15 12 28 25](https://user-images.githubusercontent.com/53387/173806440-172e6f24-5539-4338-94d7-9f995e7b97cd.gif)

To avoid this, could we either 1) disable text selection on those items, or 2) implement some custom infinite list item renderer, that makes each item draggable? I don't mean to convert the list to a draggable list, just that you could drag elements out... and drop them somewhere. I'm imagining something similar to how you'd drag a file onto a "drop your files here" area. Could that be an easy enough workaround to implement?

mariusandra commented:
> When the save is executed, we should show a toast notification that communicates the result of the action - successfully saved or error.

I also see it, though for whatever reason, I didn't earlier. Now it works well:

![2022-06-15 12 36 57](https://user-images.githubusercontent.com/53387/173807850-c4f02674-a885-4781-915e-d6f40c85fbdb.gif)

kappa90 committed:
fixes after review

kappa90 committed:
add migration

kappa90 committed:
Merge branch 'master' into save-live-events-columns

kappa90 committed:
fix black formatting

kappa90 committed:
fix cypress test

kappa90 committed:
Merge branch 'master' into save-live-events-columns

kappa90 commented:
I think we're good to go.

clarkus reviewed (changes requested):
Looks good to me. The only thing I see outstanding is items that reside in both lists. I feel like if you select an item from the available list, it should be removed once it's in the visible. list. Otherwise this looks good to me.

mariusandra commented:
I still do have these two open questions/points here  https://github.com/PostHog/posthog/pull/10293#pullrequestreview-1007243066 - would be good to get some thoughts, even if the answer is "out of scope".

Regarding removing results from the API, @kappa90 had [written before](https://github.com/PostHog/posthog/pull/10293#issuecomment-1156268414):

> > - When an available column is selected, should it be removed from the available list and then only shown in the visible list?
> 
> I spoke with @mariusandra about it and this is not so straightforward, it would require to change the taxonomic filter to call the backend requesting to exclude some results, it can't just be easily hidden on the frontend. I'd keep it out of scope.

Though looking at it, we already have this in the backend, if we pass `excluded_properties` in the request to get the property definitions:
https://github.com/PostHog/posthog/blob/5d53d76718875f17f4cfaabfc136f1be15580747/posthog/api/property_definition.py#L115

It's a bit more wiring on the frontend to set this up, and I'd happily leave it for a follow up just so we get in more chunks/PRs faster. Would be great to have it though :D.

kappa90 committed:
Merge branch 'master' into save-live-events-columns

kappa90 committed:
remove items from lists after selection

kappa90 committed:
disable text selection in infinite lists

kappa90 committed:
fix bug

kappa90 commented:
> Looks good to me. The only thing I see outstanding is items that reside in both lists. I feel like if you select an item from the available list, it should be removed once it's in the visible. list. Otherwise this looks good to me.

I fixed this. It now removed the item from the taxonomic filter list after being selected. Thanks @mariusandra for pointing out where to look, I wouldn't have found it!

> I still do have these two open questions/points here [#10293 (review)](https://github.com/PostHog/posthog/pull/10293#pullrequestreview-1007243066) - would be good to get some thoughts, even if the answer is "out of scope".

Sorry I didn't see these before. I fixed the inlining problem for truncated elements. As for drag and drop, I tried some solutions but nothing I could wrap up in a reasonable time, so I just disabled text selection on those items for now.

So, apart from the drag and drop from the right list which I'd keep out of scope for now, I think we're done?

kappa90 committed:
Merge branch 'master' into save-live-events-columns

mariusandra commented on frontend/src/lib/components/TaxonomicFilter/types.ts:
```diff
@@ -19,6 +19,8 @@ export interface TaxonomicFilterProps {
     popoverEnabled?: boolean
     selectFirstItem?: boolean
+    /** use to filter results in a group by name, currently only working for EventProperties */
+    excludedProperties?: { [key in TaxonomicFilterGroupType]?: string[] }
```
Would `Record<TaxonomicFilterGroupType, string[]>` also work? I think it'd be easier to understand.

kappa90 replied:
No, that doesn't work because it doesn't map automatically the enum keys to the record keys, unfortunately!

mariusandra replied:
Ah, got it... 😄 🤷

mariusandra commented on frontend/src/lib/components/TaxonomicFilter/infiniteListLogic.ts:
```diff
@@ -123,19 +123,26 @@ export const infiniteListLogic = kea<infiniteListLogicType>({
                         offset,
                     }
 
+                    if (excludedProperties) {
+                        searchParams = {
+                            ...searchParams,
+                            excluded_properties: JSON.stringify(excludedProperties),
+                        }
+                    }
```
It could as well be a `const` here, and you can do just `searchParams.excluded_properties = JSON.stringify(excludedProperties)`. Until we put variables into kea/redux, you can mutate them in whatever way you want to.

kappa90 replied:
Done!

mariusandra reviewed (approved):
Great work! I left two non-blocking code nits. Feel free to check or ignore and merge 👍

kappa90 committed:
fix nit

kappa90 committed:
Merge remote-tracking branch 'origin/master' into save-live-events-columns
